### ISSUE TITLE
Keras LR callbacks have unintended behavior when resuming from checkpoint

### ISSUE THREAD

**Issue by sparticlesteve** [horovod/horovod#1921]
**Environment:**
1. Framework: tf.keras
2. Framework version: TF 1.15.0
3. Horovod version: 0.19.0
4. MPI version: cray-mpich/7.7.10
5. CUDA version: n/a
6. NCCL version: n/a
7. Python version: 3.7.4
8. OS and version: Cray linux based on SLES 15
9. GCC version: 7.3.0

**Checklist:**
1. Did you search issues to find if somebody asked this question before? yes
2. If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.rst)? n/a
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.rst)? n/a
4. Did you check if you question is answered in the [troubleshooting guide](https://github.com/horovod/horovod/blob/master/docs/troubleshooting.rst)? yes

**Bug report:**
Please describe erroneous behavior you're observing and steps to reproduce it.

There is a problem with the Keras learning rate callbacks (inheriting from `LearningRateScheduleCallbackImpl`) as implemented when using checkpoints and resuming training. This class pulls the `initial_lr` from the model optimizer in `on_train_begin`:
https://github.com/horovod/horovod/blob/d1b13ec131af22b31d0ba999dae15a29991cfeae/horovod/_keras/callbacks.py#L137
All LR (and momentum) modifications are done with respect to that initial learning rate and the current epoch (or batch). However, if one is writing a checkpoint, the current modified learning rate and momentum are written to the checkpoint file. Then, upon loading that checkpoint and resuming training with the LR callback, it pulls the _modified_ LR as its new `initial_lr`. Unless the user takes care to reset the optimizer's LR (and momentum) after loading from checkpoint and before training, the original schedule applied will not produce the intended schedule. The Keras Imagenet resnet50 example is affected by this, for instance:
https://github.com/horovod/horovod/blob/master/examples/keras_imagenet_resnet50.py

In contrast, the LR scheduler in Keras (and tf.keras) is implemented such that modifications depend on the _current_ learning rate. This slightly different approach is therefore not affected by the checkpoint resume issue. It doesn't have any momentum correction, though.
https://github.com/tensorflow/tensorflow/blob/bab74a15d9ad6bb9066b3e31d601d6a45b1cb221/tensorflow/python/keras/callbacks.py#L1349

I think there are a couple of possible reasonable solutions. One is to change the multiplier logic to match that of the Keras LR scheduler so that the new LR is a result of the multiplier times the current LR. This logic change would likely break folks' code, though. Another possible solution is to allow (or require!) the user to set the initial LR in the callback constructor. This way I can ensure that it is always set to the correct, intended value. Finally, as I alluded to above, the user can reset their optimizer LR (and appropriately scale the momentum) after loading from checkpoint and before training. However, I consider this a workaround rather than a solution.

I'm hoping I explained it clearly enough that I don't need a MWE, but I can provide one if required.

**Comment by tgaddair** [comment]
Hey @sparticlesteve, thanks for the detailed report.  I am inclined towards letting the user optionally provide the `initial_lr` in the callback constructor, and only pulling it in `on_train_begin` if it hasn't been set.  That way, we don't break the existing behavior, but provide a solution for for this scenario.

Curious if other contributors have any thoughts on this.

@alsrgv @nvcastet @romerojosh

**Comment by alsrgv** [comment]
+1 for optional `initial_lr`, wouldn't want to introduce a breaking change.

**Comment by romerojosh** [comment]
Should it really be optional though? It sounds like what @sparticlesteve is describing here is erroneous behavior on restart, which really ought to be fixed. I can't imagine any user would want their LR schedule somehow dependent on if/when they restart.

**Comment by alsrgv** [comment]
Breaking existing scripts it not nice without major version bump, some people may not even be using restarts from checkpoints :-)

**Comment by tgaddair** [comment]
One option would be to add the `initial_lr` parameter as optional for now, but with a `DeprecationWarning` that we will remove the option to set `initial_lr` implicitly in a future version (e.g., `0.21.0`).  Thoughts?

**Comment by nvcastet** [comment]
@tgaddair yes, seems to be the best solution.
`tf.keras.callbacks.LearningRateScheduler` in the next TF version seems to go from LR scheduler not taking any learning rate as input (only current epoch) to taking the current epoch and current LR: https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/LearningRateScheduler?version=nightly
I am assuming customers can use `tf.keras.callbacks.LearningRateScheduler` in a Horovod script with no problem if they want a scheduler based on current learning rate, correct?

**Comment by tgaddair** [comment]
Good point, @nvcastet.  Yes, users can use the Keras LR scheduler as an alternative.

**Comment by sparticlesteve** [comment]
Yeah, I mentioned the Keras implementations above, and this is my current workaround, but as I noted, one (perhaps minor) difference is that the Keras ones do not do any momentum correction.

**Comment by tgaddair** [comment]
Hey @sparticlesteve, can you take a look at #1933 and verify that it solves the issue for you?

**Comment by sparticlesteve** [comment]
Thanks. I'll give it a shot when I have time and report back.
